File: Algorithm/BackTracking/양궁대회.swift
import Foundation

func solution(_ n:Int, _ info:[Int]) -> [Int] {
    var answer:[Int] = [-1]
    var maxDiff: Int = -1
        
    for cur in makeTotalLionCase(n, info) {
        let lion = convertArrows(cur)
        let scores = calScore(lion, info)
        
        if scores[0] <= scores[1] {
            continue
        }
        
        if maxDiff < abs(scores[0] - scores[1]) {
            maxDiff = abs(scores[0] - scores[1])
            answer = lion
        } else if maxDiff == abs(scores[0] - scores[1]) {
            answer = chechMoreLower(lion, answer)
        }
    }
    
    return answer
}

func makeTotalLionCase(_ n:Int, _ info:[Int]) -> [[Int]] {
    var result: [[Int]] = []
    
    func dfs(_ com: [Int], _ cur: Int, _ depth: Int) {
        if depth == n {
            result.append(com)
            return
        }
        
        for i in cur...10 {
            dfs(com + [i], i, depth + 1)
        }
    } // dfs
    
    dfs([], 0, 0)
    
    return result
} // makeTotalLionCase

func convertArrows(_ arr: [Int]) -> [Int] {
    var result = Array(repeating: 0, count: 11)
    
    for a in arr {
        result[a] += 1
    }
    
    return result
} // convertArrows

func calScore(_ lion: [Int], _ apach: [Int]) -> [Int] {
    var lScore: Int = 0
    var aScore: Int = 0
    
    for i in 0...10 {
        if lion[i] == 0 && apach[i] == 0 {
            continue
        }
        
        if lion[i] > apach[i] {
            lScore += 10 - i
        } else {
            aScore += 10 - i
        }
    }
    
    return [lScore, aScore]
} // check

func chechMoreLower(_ arr1: [Int], _ arr2: [Int]) -> [Int] {
    for i in stride(from: 10, through: 0, by: -1) {
        if arr1[i] < arr2[i] {
            return arr2
        } else if arr1[i] > arr2[i] {
            return arr1
        }
    }
    return arr2
} // chechMoreLower

File: Algorithm/BackTracking/양궁대회.py
from itertools import combinations_with_replacement

def solution(n, info):
    answer = []
    candi = {}
    
    for combi in combinations_with_replacement([i for i in range(11)], n):
        lion = [0] * 11
        for c in combi:
            lion[c] += 1
        
        lion_score, apach_score = get_scores(lion, info)
        
        if lion_score <= apach_score:
            continue
        
        if not lion_score - apach_score in candi:
            candi[lion_score - apach_score] = [lion]
        else:
            candi[lion_score - apach_score] += [lion]
            
            
    if not len(candi):
        answer = [-1]
    else:
        max_diff = max(candi.keys())
        answer = candi[max_diff][0]
        
        for arr in candi[max_diff][1:]:
            if better(arr, answer):
                answer = arr[:]
            
    return answer
# solution

def get_scores(lion, apach):
    lion_score, apach_score = 0, 0
    
    for i in range(11):
        if lion[i] == 0 and apach[i] == 0:
            continue
        if lion[i] > apach[i]:
            lion_score += (10 - i)
        else:
            apach_score += (10 - i)
    
    return lion_score, apach_score
# get_scores

def better(a, b):
    for i in range(10, -1, -1):
        if a[i] > b[i]:
            return True
        if a[i] < b[i]:
            return False
    return False
# better